nmigen_boards/genesys2.py

```diff
@@ -0,0 +1,390 @@
+import os
+import subprocess
+
+from nmigen.build import Resource, Pins, PinsN, DiffPairs, Subsignal, Attrs, \
+    Clock, Connector
+from nmigen.vendor.xilinx_7series import Xilinx7SeriesPlatform
+from .resources import LEDResources, ButtonResources, SwitchResources, \
+    UARTResource, SDCardResources
+
+__all__ = ["Genesys2Platform"]
+
+
+class Genesys2Platform(Xilinx7SeriesPlatform):
+    """Platform file for Diglient Genesys2 Kitex-7 board.
+    https://reference.digilentinc.com/reference/programmable-logic/genesys-2/start"""
+
+    device = "xc7k325t"
+    package = "ffg900"
+    speed = "2"
+    default_clk = "clk"
+    default_rst = "rst"
+
+    def __init__(self, JP6="2V5"):
+        super().__init__()
+
+        assert JP6 in ["1V2", "1V8", "2V5", "3V3"]
+        self._JP6 = JP6
+
+    def bank15_16_17_iostandard(self):
+        return "LVCMOS" + self._JP6
+
+    resources = [
+        *ButtonResources(pins={
+                "w": "M20",
+                "e": "C19",
+                "n": "B19",
+                "s": "M19",
+                "c": "E18"}, attrs=Attrs(IOSTANDARD=bank15_16_17_iostandard)),
+        *SwitchResources(pins="G19 G25 H24 K19 N19 P19",
+                         attrs=Attrs(IOSTANDARD=bank15_16_17_iostandard)),
+        *SwitchResources(pins={
+                6: "P26",
+                7: "P27"}, attrs=Attrs(IOSTANDARD="LVCMOS33")),
+        Resource("clk", 0, DiffPairs(p="AD12 ", n="AD11", dir="i"),
+                 Clock(200e6), Attrs(IOSTANDARD="LVDS")),
+        Resource("rst", 0, PinsN("R19", dir="i"),
+                 Attrs(IOSTANDARD="LVCMOS33")),
+        *LEDResources(pins="T28 V19 U30 U29 V20 V26 W24 W23",
+                      attrs=Attrs(IOSTANDARD="LVCMOS33")),
+        Resource("fan", 0,
+                 Subsignal("pwm", Pins("W19", dir="o")),
+                 Subsignal("tach", Pins("V21", dir="i")),
+                 Attrs(IOSTANDARD="LVCMOS33")),
+        UARTResource(0, rx="Y20", tx="AE20",
+                     attrs=Attrs(IOSTANDARD="LVCMOS33")),
+        Resource("i2c", 0,
+                 Subsignal("scl", Pins("AE30", dir="io")),
+                 Subsignal("sda", Pins("AF30", dir="io")),
+                 Attrs(IOSTANDARD="LVCMOS33")),
+        Resource("ddram", 0,
```

Comment:
This resource should be called `ddr` or `ddr2` or whatever standard it is on this board.

Reply:
It's DDR3, I also rename `cke` to `clk_en` to follow the convention. And *DDR3_RESET_N* should become a dedicated resource.

Reply:
`reset` will be renamed to `rst` as well here.

Reply:
* `ras_n` → `ras` 
* `cas_n` → `cas`